go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,321 @@
+				os.Exit(1)
+			}
+
+			err = hacluster.ConfigureAsStandby(identity)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				panic(err.Error())
+			}
+
+		}
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDeadline,
+		RetryPeriod:   retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {},
+			OnStoppedLeading: func() {
+				//This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				logrus.Infof("[%s] stopped Leading. Exiting", hacluster.nodeName)
+				os.Exit(1)
+			},
+			OnNewLeader: HAClusterNewLeader,
+		},
+	}
+
+	go leaderelection.RunOrDie(context.Background(), lec)
+
+	return nil
+}
+
+// ConfigureAsActive configures the node as active.
+func (hacluster *OvnHAMasterController) ConfigureAsActive(masterNodeName string) error {
+	if hacluster.manageDBServers {
+		// Find the endpoint for the service
+		ep, err := hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+		createEp := true
+		if err == nil {
+			if ep.Name == ovnkubeDbEp {
+				createEp = false
+			}
+		}
+
+		const (
+			nbPort int32 = 6641
+			sbPort int32 = 6642
+		)
+
+		ovndbEp := v1.Endpoints{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: config.Kubernetes.OVNConfigNamespace,
+				Name:      ovnkubeDbEp,
+			},
+			Subsets: []v1.EndpointSubset{
+				{
+					Addresses: []v1.EndpointAddress{
+						{IP: config.Kubernetes.OVNKubePodIP},
+					},
+					Ports: []v1.EndpointPort{
+						{
+							Name: "north",
+							Port: nbPort,
+						},
+						{
+							Name: "south",
+							Port: sbPort,
+						},
+					},
+				},
+			},
+		}
+
+		if createEp {
+			_, err = hacluster.cluster.Kube.CreateEndpoint(config.Kubernetes.OVNConfigNamespace, &ovndbEp)
+			if err != nil {
+				logrus.Errorf("%s Endpount Create failed", ovnkubeDbEp)
+				return err
+			}
+		} else {
+			_, err = hacluster.cluster.Kube.UpdateEndpoint(config.Kubernetes.OVNConfigNamespace, &ovndbEp)
+			if err != nil {
+				logrus.Errorf("%s Endpount Update failed", ovnkubeDbEp)
+				return err
+			}
+		}
+
+		// Promote the OVN DB servers.
+		err = hacluster.PromoteOVNDbs(strconv.Itoa(int(nbPort)), strconv.Itoa(int(sbPort)))
+		if err != nil {
+			logrus.Errorf("Promoting OVN ovsdb-servers to active failed")
+			return err
+		}
+	}
+
+	// run the cluster controller to init the master
+	err := hacluster.cluster.StartClusterMaster(hacluster.nodeName)
+	if err != nil {
+		return err
+	}
+
+	return hacluster.ovnController.Run()
+}
+
+// ConfigureAsStandby configures the node as standby
+func (hacluster *OvnHAMasterController) ConfigureAsStandby(masterNodeName string) error {
```

Comment:
You will need a persistent endpoint watch here - endpoints can change at any time, for example if pods B and C swap leadership...

Reply:
When they change leadership, the function OnNewLeader will be called right ?

I mean if we have nodes A, B and C and if B and C swap leadership, A would be notified with the 'OnNewLeader' callback right ? This is what I noticed in my testing.

Reply:
Yes, but leadership changes are different from endpoint changes. While they usually happen at the same time, there's no guaranteeing that they do.

Reply:
Ok. Sure.
Something like this ? https://github.com/numansiddique/ovn-kubernetes/blob/ha_master_support/go-controller/pkg/cluster/ha_master.go#L97 ?

Can I use the same watchFactory/channel to watch for endpoint changes ?

Reply:
That should work, yes.